src/inmanta/agent/agent.py

```diff
@@ -1033,41 +1232,52 @@ async def get_facts(self, resource: JsonType) -> Apireturn:
         self, version: int, action: const.ResourceAction, resources: list[JsonType]
-    ) -> tuple[dict[ResourceVersionIdStr, const.ResourceState], list[Resource]]:
-        """Deserialize all resources and load all handler code. When the code for this type fails to load, the resource
-        is marked as failed
+    ) -> tuple[dict[ResourceVersionIdStr, const.ResourceState], list[dict[str, object]], Executor]:
+        # TODO refactor wiring of tuples around -> not sure how
+
+        """
+        Setup an executor for resource deployment
+
+        returns a tuple
+
+            - undeployable: resources for which code loading failed
+            - loaded_resources: ALL resources from this batch
+            - the executor: with relevant handler code loaded
+
         """
         started = datetime.datetime.now().astimezone()
-        failed_resource_types = await self.process.ensure_code(
+        executor: Executor
+        failed_resource_types: set[str]
+        code: Sequence[InstallBlueprint] = await self.process.get_code(
             self._env_id, version, [res["resource_type"] for res in resources]
         )
-        loaded_resources: list[Resource] = []
+        executor, failed_resource_types = await self.get_executor(
+            code, self.get_client()
+        )  # deviation from the diagram (ie call from here instead of get_latest_version_for_agent)
+
+        loaded_resources: list[dict[str, object]] = []
         failed_resources: list[ResourceVersionIdStr] = []
         undeployable: dict[ResourceVersionIdStr, const.ResourceState] = {}
 
         for res in resources:
-            try:
-                res["attributes"]["id"] = res["id"]
-                if res["resource_type"] not in failed_resource_types:
-                    resource: Resource = Resource.deserialize(res["attributes"])
-                    loaded_resources.append(resource)
-
-                    state = const.ResourceState[res["status"]]
-                    if state in const.UNDEPLOYABLE_STATES:
-                        undeployable[res["id"]] = state
-                else:
-                    failed_resources.append(res["id"])
-                    undeployable[res["id"]] = const.ResourceState.unavailable
-                    resource = Resource.deserialize(res["attributes"], use_generic=True)
-                    loaded_resources.append(resource)
+            res["attributes"]["id"] = res["id"]
+            if res["resource_type"] not in failed_resource_types:
+                loaded_resources.append(res["attributes"])
 
-            except Exception:
+                state = const.ResourceState[res["status"]]
+                if state in const.UNDEPLOYABLE_STATES:
+                    undeployable[res["id"]] = state
+            else:
                 failed_resources.append(res["id"])
                 undeployable[res["id"]] = const.ResourceState.unavailable
-                resource = Resource.deserialize(res["attributes"], use_generic=True)
```

Comment:
If I understand correctly we only ever did this when we also added it to `undeployable`, meaning that the scheduler would never even try to deploy it, therefore we don't need the specific handler type. We did it just to present a `Resource`-like object, because that is what the surrounding code expected. Now that we use dicts here, there's no need for that anymore, and at the point where we deserialize now, there is no point to keeping the object at all anymore, we can just report to the server directly and be done.

Reply:
Makes sense :+1:

Reply:
Can you link the ticket if/when you create(d) it?